vault/identity_store_oidc.go

```diff
@@ -0,0 +1,1349 @@
+	Expiry   int64  `json:"exp"` // Expiration, as determined by the role.
+	IssuedAt int64  `json:"iat"` // Time of token creation
+}
+
+type discovery struct {
+	Issuer        string   `json:"issuer"`
+	Auth          string   `json:"authorization_endpoint"`
+	Token         string   `json:"token_endpoint"`
+	Keys          string   `json:"jwks_uri"`
+	ResponseTypes []string `json:"response_types_supported"`
+	Subjects      []string `json:"subject_types_supported"`
+	IDTokenAlgs   []string `json:"id_token_signing_alg_values_supported"`
+	Scopes        []string `json:"scopes_supported"`
+	AuthMethods   []string `json:"token_endpoint_auth_methods_supported"`
+	Claims        []string `json:"claims_supported"`
+}
+
+// oidcCache stores common response data as well as when the periodic func needs
+// to run. This is conservatively managed, and most writes to the OIDC endpoints
+// will invalidate the cache.
+type oidcCache struct {
+	lock              sync.RWMutex
+	discoveryResponse []byte
+	jwksResponse      []byte
+	config            *oidcConfig
+	nextPeriodicRun   time.Time
+}
+
+const (
+	issuerPath           = "/v1/identity/oidc"
+	oidcTokensPrefix     = "oidc_tokens/"
+	oidcConfigStorageKey = oidcTokensPrefix + "config/"
+	namedKeyConfigPath   = oidcTokensPrefix + "named_keys/"
+	publicKeysConfigPath = oidcTokensPrefix + "public_keys/"
+	roleConfigPath       = oidcTokensPrefix + "roles/"
+)
+
+var requiredClaims = []string{"iat", "aud", "exp", "iss", "sub"}
+
+func oidcPaths(i *IdentityStore) []*framework.Path {
+	return []*framework.Path{
+		{
+			Pattern: "oidc/config/?$",
+			Fields: map[string]*framework.FieldSchema{
+				"issuer": {
+					Type:        framework.TypeString,
+					Description: "Issuer URL to be used in the iss claim of the token. If not set, Vault's app_addr will be used.",
+				},
+			},
+			Callbacks: map[logical.Operation]framework.OperationFunc{
+				logical.ReadOperation:   i.pathOIDCReadConfig,
+				logical.UpdateOperation: i.pathOIDCUpdateConfig,
+			},
+			HelpSynopsis:    "OIDC configuration",
+			HelpDescription: "Update OIDC configuration in the identity backend",
+		},
+		{
+			Pattern: "oidc/key/" + framework.GenericNameRegex("name"),
+			Fields: map[string]*framework.FieldSchema{
+				"name": {
+					Type:        framework.TypeString,
+					Description: "Name of the key",
+				},
+
+				"rotation_period": {
+					Type:        framework.TypeDurationSecond,
+					Description: "How often to generate a new keypair.",
+					Default:     "6h",
+				},
+
+				"verification_ttl": {
+					Type:        framework.TypeDurationSecond,
+					Description: "Controls how long the public portion of a key will be available for verification after being rotated.",
+					Default:     "24h",
+				},
+
+				"algorithm": {
+					Type:        framework.TypeString,
+					Description: "Signing algorithm to use. This will default to RS256, and is currently the only allowed value.",
+					Default:     "RS256",
+				},
```

Comment:
Another issue with this is that RS256 just means RSA signing with SHA2-256. But it doesn't indicate anything about the size of the RSA key, which you've hardcoded. I think you need another `key_bits` field that allows the user to set the number of bits to use, with the default depending on the algorithm selected. There will be people wanting 4096 bit RSA keys, even if we don't do EC keys yet...

Reply:
Added Jira to review this.